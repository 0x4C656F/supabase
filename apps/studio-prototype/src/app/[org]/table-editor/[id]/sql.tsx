'use client'

import '@/src/styles/react-data-grid.css'
import 'react-data-grid/lib/styles.css'
import { ResizableHandle, ResizablePanel, ResizablePanelGroup } from 'ui'

import DataGrid, { Column } from 'react-data-grid'
import { Checkbox_Shadcn_ } from 'ui'
import FilterHeader from '../filter-header'
import Table from './table'
import SqlToolbar from './sql-toolbar'
import CodeSample from './code-sample'

// Define the type for your row data
interface Row {
  name: string
  company: string
  email: string
  phone: string
}

export default function Sql() {
  const rows: Row[] = [
    {
      name: 'Hines Fowler',
      company: 'BUZZNESS',
      email: 'hinesfowler@buzzness.com',
      phone: '+1 (869) 405-3127',
    },
    {
      name: 'Hines Fowler',
      company: 'BUZZNESS',
      email: 'hinesfowler@buzzness.com',
      phone: '+1 (869) 405-3127',
    },
    {
      name: 'Hines Fowler',
      company: 'BUZZNESS',
      email: 'hinesfowler@buzzness.com',
      phone: '+1 (869) 405-3127',
    },
    {
      name: 'Hines Fowler',
      company: 'BUZZNESS',
      email: 'hinesfowler@buzzness.com',
      phone: '+1 (869) 405-3127',
    },
    {
      name: 'Hines Fowler',
      company: 'BUZZNESS',
      email: 'hinesfowler@buzzness.com',
      phone: '+1 (869) 405-3127',
    },
    {
      name: 'Hines Fowler',
      company: 'BUZZNESS',
      email: 'hinesfowler@buzzness.com',
      phone: '+1 (869) 405-3127',
    },
    {
      name: 'Hines Fowler',
      company: 'BUZZNESS',
      email: 'hinesfowler@buzzness.com',
      phone: '+1 (869) 405-3127',
    },
    {
      name: 'Hines Fowler',
      company: 'BUZZNESS',
      email: 'hinesfowler@buzzness.com',
      phone: '+1 (869) 405-3127',
    },
    {
      name: 'Hines Fowler',
      company: 'BUZZNESS',
      email: 'hinesfowler@buzzness.com',
      phone: '+1 (869) 405-3127',
    },
    {
      name: 'Hines Fowler',
      company: 'BUZZNESS',
      email: 'hinesfowler@buzzness.com',
      phone: '+1 (869) 405-3127',
    },
    {
      name: 'Hines Fowler',
      company: 'BUZZNESS',
      email: 'hinesfowler@buzzness.com',
      phone: '+1 (869) 405-3127',
    },
    {
      name: 'Hines Fowler',
      company: 'BUZZNESS',
      email: 'hinesfowler@buzzness.com',
      phone: '+1 (869) 405-3127',
    },
    {
      name: 'Hines Fowler',
      company: 'BUZZNESS',
      email: 'hinesfowler@buzzness.com',
      phone: '+1 (869) 405-3127',
    },
    {
      name: 'Hines Fowler',
      company: 'BUZZNESS',
      email: 'hinesfowler@buzzness.com',
      phone: '+1 (869) 405-3127',
    },
  ]

  // Define the columns with proper types
  const columns: Column<Row>[] = [
    {
      key: 'id',
      name: '',
      width: 10,
      renderCell: ({ row }) => <Checkbox_Shadcn_ className="ml-1" />,
    },
    {
      key: 'name',
      name: 'Name',
      width: 200,
      //   renderCell: ({ row }) => <ColumnRender row={row} />,
    },
    {
      key: 'company',
      name: 'Company',
      width: 200,
      //   renderCell: ({ row }) => <ColumnRender row={row} />,
    },
    {
      key: 'email',
      name: 'Email',
      width: 200,
      //   renderCell: ({ row }) => <ColumnRender row={row} />,
    },
    {
      key: 'phone',
      name: 'Phone',
      width: 200,
      //   renderCell: ({ row }) => <ColumnRender row={row} />,
    },
  ]

  return (
    <ResizablePanelGroup direction="vertical" className="z-30">
      <ResizablePanel defaultSize={50} className="bg-surface-100">
        <CodeSample />
      </ResizablePanel>
      <ResizableHandle withHandle />
      <ResizablePanel defaultSize={50} className="flex flex-col">
        <SqlToolbar />
        <div className="grow h-full flex flex-col">
          <Table hideToolbar />
        </div>
      </ResizablePanel>
    </ResizablePanelGroup>
  )
}
