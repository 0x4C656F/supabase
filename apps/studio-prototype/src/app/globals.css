@tailwind base;
@tailwind components;
@tailwind utilities;

@import './../../../../packages/ui/build/css/source/global.css';
@import './../../../../packages/ui/build/css/themes/dark.css';
@import './../../../../packages/ui/build/css/themes/deep-dark.css';
@import './../../../../packages/ui/build/css/themes/light.css';

:root {
  --foreground-rgb: 0, 0, 0;
  --background-start-rgb: 214, 219, 220;
  --background-end-rgb: 255, 255, 255;
}

@media (prefers-color-scheme: dark) {
  :root {
    --foreground-rgb: 255, 255, 255;
    --background-start-rgb: 0, 0, 0;
    --background-end-rgb: 0, 0, 0;
  }
}

body {
  @apply bg-200;
  color: rgb(var(--foreground-rgb));
}

@layer utilities {
  .text-balance {
    text-wrap: balance;
  }
}
