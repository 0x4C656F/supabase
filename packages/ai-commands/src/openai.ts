import OpenAI from 'openai'

export const supportedModels = [
  'llama3',
  'llama3-70b-8192',
  'llama3-8b-8192',
  'gpt-4-turbo',
  'gpt-4-turbo-preview',
  'gpt-4-turbo-2024-04-09',
  'gpt-4-0125-preview',
  'gpt-4-1106-preview',
  'gpt-4',
  'gpt-4-0613',
  'gpt-4-0314',
  'gpt-3.5-turbo',
  'gpt-3.5-turbo-0125',
  'gpt-3.5-turbo-1106',
]

export const supportedEmbeddingModels = ['text-embedding-ada-002']

/**
 * Instantiates OpenAI client while checking and validating
 * environment variables.
 *
 * Allows you to use your own OpenAI-compatible provider.
 *
 * Uses the following env vars:
 * - OPENAI_API_KEY (backwards compatible with previous OPENAI_KEY)
 * - OPENAI_BASE_URL
 * - OPENAI_MODEL
 */
export function createOpenAiClient() {
  const apiKey = process.env.OPENAI_API_KEY ?? process.env.OPENAI_KEY
  const baseURL = process.env.OPENAI_BASE_URL
  const model = process.env.OPENAI_MODEL

  if (!apiKey) {
    return {
      error: new Error(
        'No OPENAI_API_KEY set. Create this environment variable to use AI features.'
      ),
    }
  }

  if (!baseURL) {
    return {
      error: new Error(
        'No OPENAI_BASE_URL set. Create this environment variable to use AI features.'
      ),
    }
  }

  if (!model) {
    return {
      error: new Error('No OPENAI_MODEL set. Create this environment variable to use AI features.'),
    }
  }

  if (!supportedModels.includes(model)) {
    return {
      error: new Error(
        `The following models are supported for OPENAI_MODEL: ${JSON.stringify(supportedModels)}`
      ),
    }
  }

  const openai = new OpenAI({ apiKey, baseURL })

  return { openai, model }
}

/**
 * Instantiates OpenAI client for embedding models while checking and validating
 * environment variables.
 *
 * Allows you to use separate OpenAI-compatible providers for embedding models vs. language models.
 *
 * Uses the following env vars:
 * - OPENAI_EMBEDDING_API_KEY
 * - OPENAI_EMBEDDING_BASE_URL
 * - OPENAI_EMBEDDING_MODEL
 *
 * Falls back to these default env vars if the above vars don't exist:
 * - OPENAI_API_KEY
 * - OPENAI_BASE_URL
 */
export function createOpenAiEmbeddingClient() {
  const apiKey =
    process.env.OPENAI_EMBEDDING_API_KEY ?? process.env.OPENAI_API_KEY ?? process.env.OPENAI_KEY
  const baseURL = process.env.OPENAI_EMBEDDING_BASE_URL ?? process.env.OPENAI_BASE_URL
  const model = process.env.OPENAI_EMBEDDING_MODEL

  if (!apiKey) {
    return {
      error: new Error(
        'No OPENAI_EMBEDDING_API_KEY set. Create this environment variable to use AI embedding features.'
      ),
    }
  }

  if (!baseURL) {
    return {
      error: new Error(
        'No OPENAI_EMBEDDING_BASE_URL set. Create this environment variable to use AI embedding features.'
      ),
    }
  }

  if (!model) {
    return {
      error: new Error(
        'No OPENAI_EMBEDDING_MODEL set. Create this environment variable to use AI embedding features.'
      ),
    }
  }

  if (!supportedEmbeddingModels.includes(model)) {
    return {
      error: new Error(
        `The following models are supported for OPENAI_EMBEDDING_MODEL: ${JSON.stringify(supportedEmbeddingModels)}`
      ),
    }
  }

  const openai = new OpenAI({ apiKey, baseURL })

  return { openai, model }
}
