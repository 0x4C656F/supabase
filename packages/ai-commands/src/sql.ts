import { SchemaBuilder } from '@serafin/schema-builder'
import { codeBlock, stripIndent } from 'common-tags'
import { jsonrepair } from 'jsonrepair'
import OpenAI from 'openai'
import { ContextLengthError, EmptyResponseError, EmptySqlError } from './errors'

// Declare JSON schema for each function that the LLM can call
const generateSqlSchema = SchemaBuilder.emptySchema()
  .addString('sql', {
    description: stripIndent`
      The generated SQL (must be valid SQL).
    `,
  })
  .addString('title', {
    description: stripIndent`
      The title of the SQL.
      - Omit words like 'SQL', 'Postgres', or 'Query'
    `,
  })

const editSqlSchema = SchemaBuilder.emptySchema().addString('sql', {
  description: stripIndent`
      The modified SQL (must be valid SQL).
    `,
})

const debugSqlSchema = SchemaBuilder.emptySchema()
  .addString('solution', {
    description: 'A short suggested solution for the error (as concise as possible).',
  })
  .addString('sql', {
    description:
      'The SQL rewritten to apply the solution. Includes all the original logic, but modified to fix the issue.',
  })

const generateTitleSchema = SchemaBuilder.emptySchema()
  .addString('title', {
    description: stripIndent`
      The generated title for the SQL snippet (short and concise).
      - Omit these words: 'SQL', 'Postgres', 'Query', 'Database'
    `,
  })
  .addString('description', {
    description: stripIndent`
      The generated description for the SQL snippet.
    `,
  })

// Reference auto-generated types for each JSON schema
export type GenerateSqlResult = typeof generateSqlSchema.T
export type EditSqlResult = typeof editSqlSchema.T
export type DebugSqlResult = typeof debugSqlSchema.T
export type GenerateTitleResult = typeof generateTitleSchema.T

// Combine the completion functions
const completionFunctions = {
  generateSql: {
    name: 'generateSql',
    description: 'Generates Postgres SQL based on a natural language prompt',
    parameters: generateSqlSchema.schema as Record<string, unknown>,
  },
  editSql: {
    name: 'editSql',
    description: "Edits a Postgres SQL query based on the user's instructions",
    parameters: editSqlSchema.schema as Record<string, unknown>,
  },
  debugSql: {
    name: 'debugSql',
    description: 'Debugs a Postgres SQL error. Returns the fixed SQL and a solution explaining it.',
    parameters: debugSqlSchema.schema as Record<string, unknown>,
  },
  generateTitle: {
    name: 'generateTitle',
    description: 'Generates a short title and summarized description for a Postgres SQL snippet.',
    parameters: generateTitleSchema.schema as Record<string, unknown>,
  },
} satisfies Record<string, OpenAI.Chat.Completions.ChatCompletionCreateParams.Function>

/**
 * Generates a SQL snippet based on the provided prompt.
 *
 * @returns The generated SQL along with a title for it.
 */
export async function generateSql(
  openai: OpenAI,
  model: string,
  prompt: string,
  entityDefinitions?: string[]
) {
  const hasEntityDefinitions = entityDefinitions !== undefined && entityDefinitions.length > 0

  const completionMessages: OpenAI.Chat.Completions.ChatCompletionMessageParam[] = []

  completionMessages.push({
    role: 'system',
    content: stripIndent`
      You are a Postgres SQL generator. You must generate Postgres SQL based on a natural language prompt

      The generated SQL (must be valid SQL).
        - For primary keys, always use "id bigint primary key generated always as identity" (not serial)
        - Prefer creating foreign key references in the create statement
        - When creating new tables, always add foreign key references inline (ie. "my_column my_type references ..." on a single line)
        - Include all necessary tables and create them in the correct order (eg. book references author, so create author before book)
        - Prefer 'text' over 'varchar'
        - Prefer 'timestamp with time zone' over 'date'
        - Use vector(384) data type for any embedding/vector related query
        - Always use double apostrophe in SQL strings (eg. 'Night''s watch')
        - Always use semicolons

        Response must contain only JSON satisfying the 'tool_calls'.
    `,
  })

  if (hasEntityDefinitions) {
    completionMessages.push({
      role: 'user',
      content: codeBlock`
        Here is my database schema for reference:
        ${entityDefinitions.join('\n\n')}
      `,
    })
  }

  completionMessages.push({
    role: 'user',
    content: prompt,
  })

  try {
    const completionResponse = await openai.chat.completions.create({
      model,
      messages: completionMessages,
      max_tokens: 1024,
      temperature: 0,
      tool_choice: {
        type: 'function',
        function: {
          name: completionFunctions.generateSql.name,
        },
      },
      tools: [
        {
          type: 'function',
          function: completionFunctions.generateSql,
        },
      ],
      stream: false,
    })

    const [firstChoice] = completionResponse.choices
    const [firstTool] = firstChoice.message?.tool_calls ?? []

    const sqlResponseString = firstTool?.function.arguments

    if (!sqlResponseString) {
      throw new EmptyResponseError()
    }

    const repairedJsonString = jsonrepair(sqlResponseString)

    const result: GenerateSqlResult = JSON.parse(repairedJsonString)

    if (!result.sql) {
      throw new EmptySqlError()
    }

    return result
  } catch (error) {
    if (error instanceof Error && 'code' in error && error.code === 'context_length_exceeded') {
      throw new ContextLengthError()
    }
    throw error
  }
}

/**
 * Modifies a SQL snippet based on the provided prompt.
 *
 * @returns The modified SQL.
 */
export async function editSql(
  openai: OpenAI,
  model: string,
  prompt: string,
  sql: string,
  entityDefinitions?: string[]
) {
  const hasEntityDefinitions = entityDefinitions !== undefined && entityDefinitions.length > 0

  const completionMessages: OpenAI.Chat.Completions.ChatCompletionMessageParam[] = []

  completionMessages.push({
    role: 'system',
    content: stripIndent`
      You are a Postgres SQL editor. You must edit Postgres SQL based on the user's instructions.

      - Assume the query hasn't been executed yet
      - For primary keys, always use "id bigint primary key generated always as identity" (not serial)
      - When creating new tables, always add foreign key references inline (ie. "my_column my_type references ..." on a single line)
      - Include all necessary tables and create them in the correct order (eg. book references author, so create author before book)
      - Prefer 'text' over 'varchar'
      - Prefer 'timestamp with time zone' over 'date'
      - Use vector(384) data type for any embedding/vector related query
      - Always use double apostrophe in SQL strings (eg. 'Night''s watch')
      - Use real examples when possible
      - Add constraints if requested
      - Always use semicolons

      Response must contain only JSON satisfying the 'tool_calls'.
    `,
  })

  if (hasEntityDefinitions) {
    completionMessages.push({
      role: 'user',
      content: codeBlock`
        Here is my database schema for reference:
        ${entityDefinitions.join('\n\n')}
      `,
    })
  }

  completionMessages.push(
    {
      role: 'user',
      content: stripIndent`
        Here is my current SQL:
        ${sql}
      `,
    },
    {
      role: 'user',
      content: prompt,
    }
  )

  try {
    const completionResponse = await openai.chat.completions.create({
      model,
      messages: completionMessages,
      max_tokens: 2048,
      temperature: 0,
      tool_choice: {
        type: 'function',
        function: {
          name: completionFunctions.editSql.name,
        },
      },
      tools: [
        {
          type: 'function',
          function: completionFunctions.editSql,
        },
      ],
      stream: false,
    })

    const [firstChoice] = completionResponse.choices
    const [firstTool] = firstChoice.message?.tool_calls ?? []

    const sqlResponseString = firstTool?.function.arguments

    if (!sqlResponseString) {
      throw new EmptyResponseError()
    }

    const repairedJsonString = jsonrepair(sqlResponseString)

    const result: EditSqlResult = JSON.parse(repairedJsonString)

    if (!result.sql) {
      throw new EmptySqlError()
    }

    return result
  } catch (error) {
    if (error instanceof Error && 'code' in error && error.code === 'context_length_exceeded') {
      throw new ContextLengthError()
    }
    throw error
  }
}

/**
 * Debugs SQL errors.
 *
 * @returns A suggested SQL fix along with a solution explanation.
 */
export async function debugSql(
  openai: OpenAI,
  model: string,
  errorMessage: string,
  sql: string,
  entityDefinitions?: string[]
) {
  const hasEntityDefinitions = entityDefinitions !== undefined && entityDefinitions.length > 0

  const completionMessages: OpenAI.Chat.Completions.ChatCompletionMessageParam[] = []

  completionMessages.push({
    role: 'system',
    content: stripIndent`
      You are a Postgres SQL debugger. You must fix the user's SQL and a provide solution explaining it.

      - Create extensions if they are missing (only for valid extensions)
      - Suggest creating tables if they are missing
      - Include all of the original SQL
      - For primary keys, always use "id bigint primary key generated always as identity" (not serial)
      - When creating tables, always add foreign key references inline
      - Prefer 'text' over 'varchar'
      - Prefer 'timestamp with time zone' over 'date'
      - Use vector(384) data type for any embedding/vector related query
      - Always use double apostrophe in SQL strings (eg. 'Night''s watch')
      - Always use semicolons
      - Do not mix up the user's error with "tool use failed"
    `,
  })

  if (hasEntityDefinitions) {
    completionMessages.push({
      role: 'user',
      content: codeBlock`
        Here is my database schema for reference:
        ${entityDefinitions.join('\n\n')}
      `,
    })
  }

  completionMessages.push(
    {
      role: 'user',
      content: stripIndent`
        Here is my current SQL:
        ${sql}
      `,
    },
    {
      role: 'user',
      content: stripIndent`
        Here is the error I am getting:
        ${errorMessage}
      `,
    }
  )

  try {
    const completionResponse = await openai.chat.completions.create({
      model,
      messages: completionMessages,
      max_tokens: 2048,
      temperature: 0,
      tool_choice: {
        type: 'function',
        function: {
          name: completionFunctions.debugSql.name,
        },
      },
      tools: [
        {
          type: 'function',
          function: completionFunctions.debugSql,
        },
      ],
      stream: false,
    })

    const [firstChoice] = completionResponse.choices
    const [firstTool] = firstChoice.message?.tool_calls ?? []

    const sqlResponseString = firstTool?.function.arguments

    if (!sqlResponseString) {
      throw new EmptyResponseError()
    }

    const repairedJsonString = jsonrepair(sqlResponseString)

    const result: DebugSqlResult = JSON.parse(repairedJsonString)

    if (!result.sql) {
      throw new EmptySqlError()
    }

    return result
  } catch (error) {
    if (error instanceof Error && 'code' in error && error.code === 'context_length_exceeded') {
      throw new ContextLengthError()
    }
    throw error
  }
}

/**
 * Generates a snippet title based on the provided SQL.
 *
 * @returns A title and description for the SQL snippet.
 */
export async function titleSql(openai: OpenAI, model: string, sql: string) {
  const completionMessages: OpenAI.Chat.Completions.ChatCompletionMessageParam[] = []

  completionMessages.push({
    role: 'system',
    content: stripIndent`
      You are a Postgres SQL title/description generator.
      You must generates a short title and summarized description for a Postgres SQL snippet.

      - The title should be short but still address all parts of the code.
      - The description should describe why this table was created (eg. "Table to track todos")
      while addressing all parts of the code.
    `,
  })

  completionMessages.push({
    role: 'user',
    content: sql,
  })

  try {
    const completionResponse = await openai.chat.completions.create({
      model,
      messages: completionMessages,
      max_tokens: 1024,
      temperature: 0,
      tool_choice: {
        type: 'function',
        function: {
          name: completionFunctions.generateTitle.name,
        },
      },
      tools: [
        {
          type: 'function',
          function: completionFunctions.generateTitle,
        },
      ],
      stream: false,
    })

    const [firstChoice] = completionResponse.choices
    const [firstTool] = firstChoice.message?.tool_calls ?? []

    const sqlResponseString = firstTool?.function.arguments

    if (!sqlResponseString) {
      throw new EmptyResponseError()
    }

    const repairedJsonString = jsonrepair(sqlResponseString)

    const result: GenerateTitleResult = JSON.parse(repairedJsonString)

    return result
  } catch (error) {
    if (error instanceof Error && 'code' in error && error.code === 'context_length_exceeded') {
      throw new ContextLengthError()
    }
    throw error
  }
}
