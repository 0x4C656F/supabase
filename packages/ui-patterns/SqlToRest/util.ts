import {
  ColumnFilter,
  ColumnTarget,
  Filter,
  HttpRequest,
  Statement,
  SupabaseJsQuery,
  Target,
} from 'sql-to-rest'

export type BaseResult = {
  statement: Statement
}

export type HttpResult = BaseResult &
  HttpRequest & {
    type: 'http'
    language: 'http' | 'curl'
  }

export type SupabaseJsResult = BaseResult &
  SupabaseJsQuery & {
    type: 'supabase-js'
    language: 'js'
  }

export type ResultBundle = HttpResult | SupabaseJsResult

/**
 * Recursively iterates through PostgREST filters and checks if the predicate
 * matches any of them (ie. `some()`).
 */
export function someFilter(filter: Filter, predicate: (filter: ColumnFilter) => boolean): boolean {
  const { type } = filter

  if (type === 'column') {
    return predicate(filter)
  } else if (type === 'logical') {
    return filter.values.some((f) => someFilter(f, predicate))
  } else {
    throw new Error(`Unknown filter type '${type}'`)
  }
}

/**
 * Recursively iterates through a PostgREST target list and checks if the predicate
 * matches any of them (ie. `some()`).
 */
export function someTarget(target: Target, predicate: (target: ColumnTarget) => boolean): boolean {
  const { type } = target

  if (type === 'column-target') {
    return predicate(target)
  } else if (type === 'embedded-target') {
    return target.targets.some((t) => someTarget(t, predicate))
  } else {
    throw new Error(`Unknown target type '${type}'`)
  }
}

/**
 * Recursively flattens PostgREST embedded targets.
 */
export function flattenTargets(targets: Target[]): Target[] {
  return targets.flatMap((target) => {
    const { type } = target
    if (type === 'column-target') {
      return target
    } else if (type === 'embedded-target') {
      return [target, ...flattenTargets(target.targets)]
    } else {
      throw new Error(`Unknown target type '${type}'`)
    }
  })
}
