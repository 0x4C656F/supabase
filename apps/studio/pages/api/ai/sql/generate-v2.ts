import { StreamingTextResponse } from 'ai'
import { generateV2 } from 'ai-commands/edge'
import { createOpenAiClient } from 'ai-commands/src/openai'
import { NextRequest } from 'next/server'

export const runtime = 'edge'

export default async function handler(req: NextRequest) {
  const { method } = req

  switch (method) {
    case 'POST':
      return handlePost(req)
    default:
      return new Response(
        JSON.stringify({ data: null, error: { message: `Method ${method} Not Allowed` } }),
        {
          status: 405,
          headers: { 'Content-Type': 'application/json', Allow: 'POST' },
        }
      )
  }
}

async function handlePost(request: NextRequest) {
  const { openai, model, error } = createOpenAiClient()

  if (error) {
    return new Response(
      JSON.stringify({
        error: error.message,
      }),
      {
        status: 500,
        headers: { 'Content-Type': 'application/json' },
      }
    )
  }

  const body = await (request.json() as Promise<{
    messages: { content: string; role: 'user' | 'assistant' }[]
    existingSql?: string
    entityDefinitions: string[]
  }>)

  const { messages, existingSql, entityDefinitions } = body

  try {
    const stream = await generateV2(openai, model, messages, existingSql, entityDefinitions)
    return new StreamingTextResponse(stream)
  } catch (error) {
    if (error instanceof Error) {
      console.error(`AI SQL generation-v2 failed: ${error.message}`)
    } else {
      console.error(`AI SQL generation-v2 failed: ${error}`)
    }

    return new Response(
      JSON.stringify({
        error: 'There was an error processing your request',
      }),
      {
        status: 500,
        headers: { 'Content-Type': 'application/json' },
      }
    )
  }
}
