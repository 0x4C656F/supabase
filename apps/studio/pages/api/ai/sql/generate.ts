import { ContextLengthError, EmptySqlError, generateSql } from 'ai-commands'
import { createOpenAiClient } from 'ai-commands/src/openai'
import apiWrapper from 'lib/api/apiWrapper'
import { NextApiRequest, NextApiResponse } from 'next'

async function handler(req: NextApiRequest, res: NextApiResponse) {
  const { method } = req

  switch (method) {
    case 'POST':
      return handlePost(req, res)
    default:
      res.setHeader('Allow', ['POST'])
      res.status(405).json({ data: null, error: { message: `Method ${method} Not Allowed` } })
  }
}

export async function handlePost(req: NextApiRequest, res: NextApiResponse) {
  const { openai, model, error } = createOpenAiClient()

  if (error) {
    return res.status(500).json({ error: error.message })
  }

  const {
    body: { prompt, entityDefinitions },
  } = req

  try {
    const result = await generateSql(openai, model, prompt, entityDefinitions)
    return res.json(result)
  } catch (error) {
    if (error instanceof Error) {
      console.error(`AI SQL generation failed: ${error.message}`)

      const hasEntityDefinitions = entityDefinitions !== undefined && entityDefinitions.length > 0

      if (error instanceof ContextLengthError && hasEntityDefinitions) {
        return res.status(400).json({
          error:
            'Your database metadata is too large for Supabase AI to ingest. Try disabling database metadata in AI settings.',
        })
      }

      if (error instanceof EmptySqlError) {
        res.status(400).json({
          error: 'Unable to generate SQL. Try adding more details to your prompt.',
        })
      }
    } else {
      console.log(`Unknown error: ${error}`)
    }

    return res.status(500).json({
      error: 'There was an unknown error generating the SQL snippet. Please try again.',
    })
  }
}

const wrapper = (req: NextApiRequest, res: NextApiResponse) => apiWrapper(req, res, handler)

export default wrapper
