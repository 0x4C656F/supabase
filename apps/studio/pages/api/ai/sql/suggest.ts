import { StreamingTextResponse } from 'ai'
import { chatRlsPolicy } from 'ai-commands/edge'
import { createOpenAiClient } from 'ai-commands/src/openai'
import { NextRequest } from 'next/server'

export const runtime = 'edge'

export default async function handler(req: NextRequest) {
  const { method } = req

  switch (method) {
    case 'POST':
      return handlePost(req)
    default:
      return new Response(
        JSON.stringify({ data: null, error: { message: `Method ${method} Not Allowed` } }),
        {
          status: 405,
          headers: { 'Content-Type': 'application/json', Allow: 'POST' },
        }
      )
  }
}

async function handlePost(request: NextRequest) {
  const { openai, model, error } = createOpenAiClient()

  if (error) {
    return new Response(
      JSON.stringify({
        error: error.message,
      }),
      {
        status: 500,
        headers: { 'Content-Type': 'application/json' },
      }
    )
  }

  const body = await (request.json() as Promise<{
    messages: { content: string; role: 'user' | 'assistant' }[]
    entityDefinitions: string[]
    policyDefinition: string
  }>)

  const { messages, entityDefinitions, policyDefinition } = body

  try {
    const stream = await chatRlsPolicy(openai, model, messages, entityDefinitions, policyDefinition)
    return new StreamingTextResponse(stream)
  } catch (error) {
    console.error(error)

    return new Response(
      JSON.stringify({
        error: 'There was an error processing your request',
      }),
      {
        status: 500,
        headers: { 'Content-Type': 'application/json' },
      }
    )
  }
}
