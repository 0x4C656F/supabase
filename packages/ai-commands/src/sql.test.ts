import { describe, expect, test } from '@jest/globals'
import { codeBlock } from 'common-tags'
import { collectStream, extractMarkdownSql, formatSql, getPolicyInfo } from '../test/util'
import { createOpenAiClient } from './openai'
import { debugSql, editSql, generateSql, titleSql } from './sql'
import { chatRlsPolicy, generateV2 } from './sql.edge'

const { openai, model, error } = createOpenAiClient()

if (error) {
  throw error
}

console.info(`Running AI tests against model '${model}'`)

describe('generate', () => {
  test('single table with specified columns', async () => {
    const { sql, title } = await generateSql(
      openai,
      model,
      'create a table to track employees with name, email, and position'
    )

    expect(formatSql(sql)).toMatchSnapshot()
    await expect(title).toMatchCriteria('relates to employees')
  })
})

describe('edit', () => {
  test('add length constraint', async () => {
    const { sql } = await editSql(
      openai,
      model,
      'force name to be at least 4 characters',
      codeBlock`
        create table employees (
          id bigint primary key generated always as identity,
          name text,
          email text
        );
      `
    )

    expect(formatSql(sql)).toMatchSnapshot()
  })
})

describe('debug', () => {
  test('fix order of operations', async () => {
    const { sql } = await debugSql(
      openai,
      model,
      'relation "departments" does not exist',
      codeBlock`
        create table employees (
          id bigint primary key generated always as identity,
          name text,
          email text,
          department_id bigint references departments (id)
        );
        
        create table departments (
          id bigint primary key generated always as identity,
          name text
        );
      `
    )

    expect(formatSql(sql)).toMatchSnapshot()
  })

  test('fix typos', async () => {
    const { sql, solution } = await debugSql(
      openai,
      model,
      'syntax error at or near "fromm"',
      codeBlock`
        select * fromm employees;
      `
    )

    expect(solution).toBeDefined()
    expect(formatSql(sql)).toMatchSnapshot()
  })
})

describe('title', () => {
  test('title matches content', async () => {
    const { title, description } = await titleSql(
      openai,
      model,
      codeBlock`
        create table employees (
          id bigint primary key generated always as identity,
          name text,
          email text,
          department_id bigint references departments (id)
        );
        
        create table departments (
          id bigint primary key generated always as identity,
          name text
        );
      `
    )

    await expect(title).toMatchCriteria('relates to employees and departments')
    await expect(description).toMatchCriteria('describes employees and departments')
  })
})

describe('rls chat', () => {
  test('select policy using table definition', async () => {
    const responseStream = await chatRlsPolicy(
      openai,
      model,
      [
        {
          role: 'user',
          content: 'Users can only select their own todos',
        },
      ],
      [
        codeBlock`
          create table todos (
            id bigint primary key generated always as identity,
            task text,
            email text,
            user_id uuid references auth.users (id)
          );
        `,
      ]
    )
    const responseText = await collectStream(responseStream)
    const [sql] = extractMarkdownSql(responseText)

    const { name, ...otherInfo } = await getPolicyInfo(sql)

    expect(otherInfo).toMatchSnapshot()
    await expect(name).toMatchCriteria(
      'policy says that users can select their own todos (not insert, update, delete, etc)'
    )
  })
})

describe('sql chat', () => {
  test('single table with specified columns', async () => {
    const responseStream = await generateV2(openai, model, [
      {
        role: 'user',
        content: 'create a table to track employees with name, email, and position',
      },
    ])
    const responseText = await collectStream(responseStream)
    const [sql] = extractMarkdownSql(responseText)

    expect(sql).toMatchSnapshot()
  })

  test('table referencing users creates fk to auth.users', async () => {
    const responseStream = await generateV2(openai, model, [
      {
        role: 'user',
        content: 'create a todos table and track author',
      },
    ])

    const responseText = await collectStream(responseStream)
    const [sql] = extractMarkdownSql(responseText)

    expect(sql).toMatchSnapshot()
  })
})
