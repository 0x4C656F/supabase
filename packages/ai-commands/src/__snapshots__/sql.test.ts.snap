// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`debug fix order of operations 1`] = `
"create table departments (
  id bigint primary key generated always as identity,
  name text
);

create table employees (
  id bigint primary key generated always as identity,
  name text,
  email text,
  department_id bigint references departments (id)
);"
`;

exports[`debug fix typos 1`] = `
"select
  *
from
  employees;"
`;

exports[`edit add length constraint 1`] = `
"create table employees (
  id bigint primary key generated always as identity,
  name text check (length(name) >= 4),
  email text
);"
`;

exports[`generate single table with specified columns 1`] = `
"create table employees (
  id bigint primary key generated always as identity,
  name text not null,
  email text unique not null,
  position text not null
);"
`;

exports[`rls chat select policy using table definition 1`] = `
{
  "checkExpression": undefined,
  "command": "select",
  "table": "todos",
  "usingExpression": "user_id = auth.uid ()",
}
`;

exports[`sql chat single table with specified columns 1`] = `
"create table employees (
  id bigint primary key generated always as identity,
  name text not null,
  email text unique not null,
  position text not null,
  created_at timestamp with time zone default current_timestamp,
  updated_at timestamp with time zone default current_timestamp
);"
`;

exports[`sql chat table referencing users creates fk to auth.users 1`] = `
"create table todos (
  id bigint primary key generated always as identity,
  title text not null,
  description text,
  created_at timestamp with time zone default current_timestamp,
  author_id uuid references auth.users (id),
  completed boolean default false
);"
`;
